From 2b9b3f013467765219c8ab8a50943d3c7db68f75 Mon Sep 17 00:00:00 2001
From: Mengdong Lin <mengdong.lin@linux.intel.com>
Date: Wed, 18 Jan 2017 11:52:35 +0800
Subject: [PATCH] ucm: Assure the user input card name not to exceed max size
 of card long name

Users can load a card's UCM configuration file by giving the card short
name or long name, which should not exceed the maximum card long name
defined by the kernel. The kernel uses an 80-character buffer to store
the card long name.

Signed-off-by: Mengdong Lin <mengdong.lin@linux.intel.com>
Signed-off-by: Takashi Iwai <tiwai@suse.de>
---
 src/ucm/parser.c    | 6 ++++++
 src/ucm/ucm_local.h | 1 +
 2 files changed, 7 insertions(+)

diff --git a/src/ucm/parser.c b/src/ucm/parser.c
index f520abc5..f61201a6 100644
--- a/src/ucm/parser.c
+++ b/src/ucm/parser.c
@@ -1335,6 +1335,12 @@ static int load_master_config(const char *card_name, snd_config_t **cfg)
 	char *env = getenv(ALSA_CONFIG_UCM_VAR);
 	int err;
 
+	if (strnlen(card_name, MAX_CARD_LONG_NAME) == MAX_CARD_LONG_NAME) {
+		uc_error("error: invalid card name %s (at most %d chars)\n",
+			 card_name, MAX_CARD_LONG_NAME - 1);
+		return -EINVAL;
+	}
+
 	snprintf(filename, sizeof(filename)-1,
 		"%s/%s/%s.conf", env ? env : ALSA_USE_CASE_DIR,
 		card_name, card_name);
diff --git a/src/ucm/ucm_local.h b/src/ucm/ucm_local.h
index 6d3302f9..e41aafa4 100644
--- a/src/ucm/ucm_local.h
+++ b/src/ucm/ucm_local.h
@@ -41,6 +41,7 @@
 #include "use-case.h"
 
 #define MAX_FILE		256
+#define MAX_CARD_LONG_NAME	80
 #define ALSA_USE_CASE_DIR	ALSA_CONFIG_DIR "/ucm"
 
 #define SEQUENCE_ELEMENT_TYPE_CDEV	1
-- 
2.17.0

